appengine/genetics/js/cage.js

```diff
@@ -0,0 +1,725 @@
+  // Assign a unique id to the player and add to the game.
+  var id = Genetics.Cage.nextAvailablePlayerId_++;
+  Genetics.Cage.players[id] = [playerName, code];
+};
+
+/**
+ * Start the Cage simulation. Players should be already added
+ * @param {Function} doneCallback Function to call when game ends.
+ */
+Genetics.Cage.start = function(doneCallback) {
+  Genetics.Cage.doneCallback_ = doneCallback;
+  // Create a mouse for each player.
+  for (var playerId in Genetics.Cage.players) {
+    if (Genetics.Cage.players.hasOwnProperty(playerId)) {
+      for (var i = 0; i < Genetics.Cage.START_MICE_PER_PLAYER; i++) {
+        var mouseSex = i % 2 == 0 ? Genetics.Mouse.Sex.MALE :
+            Genetics.Mouse.Sex.FEMALE;
+        var mouseId = Genetics.Cage.nextAvailableMouseId_++;
+        var mouse = new Genetics.Mouse(mouseId, mouseSex, null, null, playerId);
+        Genetics.Cage.born(mouse);
+        new Genetics.Cage.Event('ADD', mouse,
+            Genetics.Cage.players[playerId][0]).addToQueue();
+      }
+    }
+  }
+  Genetics.Cage.endTime_ = Date.now() + Genetics.Cage.GAME_TIME_LIMIT_MSC;
+  new Genetics.Cage.Event('START_GAME').addToQueue();
+};
+
+/**
+ * Queues life simulation for mouse.
+ * @param {Genetics.Mouse} mouse The mouse to queue the simulation for.
+ */
+Genetics.Cage.life = function(mouse) {
+  // Queue life simulation for mouse with some variation in timeout to allow for
+  // variation in order.
+  Genetics.Cage.lifePidsMap_[mouse.id] =
+      setTimeout(goog.partial(Genetics.Cage.simulateLife, mouse),
+          10 * Math.random());
+};
+
+/**
+ * Runs life simulation on mouse.
+ * @param {Genetics.Mouse} mouse The mouse to run the simulation for.
+ */
+Genetics.Cage.simulateLife = function(mouse) {
+  // Remove life from mapping of lives that are queued to run.
+  delete Genetics.Cage.lifePidsMap_[mouse.id];
+  // Age mouse.
+  mouse.age++;
+  if (mouse.aggressiveness > 0) {
+    // Mouse has fight attempts left, try to start a fight with another mouse
+    Genetics.Cage.instigateFight(mouse);
+  } else if (mouse.fertility > 0) {
+    // Mouse is still fertile, try to breed with another mouse.
+    Genetics.Cage.tryMate(mouse);
+  } else {
+    // Mouse has completed life cycle and dies.
+    new Genetics.Cage.Event('RETIRE', mouse.id).addToQueue();
+    Genetics.Cage.death(mouse);
+  }
+  // Queue life simulation again if mouse is still alive.
+  if (Genetics.Cage.isAlive(mouse)) {
+    Genetics.Cage.life(mouse);
+  }
+  // Check for end game.
+  Genetics.Cage.checkForEnd();
+};
+
+/**
+ * Uses up up agression for mouse and simulates a fight between the given mouse
+ * and opponent it requests. Any mouse that loses in a fight dies.
+ * @param {Genetics.Mouse} mouse The mouse to initiate fighting.
+ */
+Genetics.Cage.instigateFight = function(mouse) {
+  var result = Genetics.Cage.runMouseFunction(mouse, 'pickFight');
+  // Check if function threw an error or caused a timeout.
+  if (result[0] == 'FAILURE') {
```

Comment:
Hard to tell what [0] is. Suggest that the return value be an object with named properties, similar to my comment for `players` above.